### ISSUE TITLE
[SPARK-39319][CORE][SQL] Make query contexts as a part of `SparkThrowable`

### ISSUE THREAD

**Issue by amahussein** [NVIDIA/spark-rapids#6182]
#### Spark-context

- https://github.com/apache/spark/commit/e9eb28e27d
- Followup: https://github.com/apache/spark/commit/90d83479a1

#### What changes were proposed in SPARK pull request?

- A new interface `QueryContext` Spark core is added that allows to get an instance of `QueryContext` from Spark's exceptions of the type `SparkThrowable`. For instance, `QueryContext` should help users to figure out where an error occur while executing queries in Spark SQL.
- Adds `SqlQueryContext` as one of implementation of `QueryContext` to Spark SQL `Origin` which contains a context of `TreeNodes` + textual summary of the error.

Example of error output:

```sql
> SELECT * FROM v1;

{
  “errorClass” : [ “DIVIDE_BY_ZERO” ],
  “parameters” : [ {
                   “name” = “config”,
                   “value” = “spark.sql.ansi.enabled”
                 }
                 ],
  “sqlState” : “42000”,
  “context” : {
      “objectType” : “VIEW”,
      “objectName” : “default.v1”
      “indexStart” : 36,
      “indexEnd” : 41,
      “fragment” : “1 / 0” }
   }
}
```

#### Why are the changes needed in Spark?

Improves the information of the error message.

#### Does this PR introduce _any_ user-facing change?

- Yes, including code interface change and output format changes

#### Why it might affect RAPIDS?

- Build will fail due to type errors.
- Apparently Spark's exceptions are modified to accept the optional `QueryContext` and pre-built text summary. The plugin will fail because it uses `QueryExecutionErrors` passing a String. We need to replace the type of `queryContext` from `String` to `Option[QueryContext]`.
- Modifications to SQL expressions init and pass new context to exceptions.

#### Impact on Testing?

Yes. May need fix to the exception messages

#### Requires Doc update?

No.

**Comment by NVnavkumar** [comment]
May have already been resolved based on Databricks 11.3 shim, and compile error fix #8129

**Comment by NVnavkumar** [comment]
This was fixed for the Databricks 11.3 shim, and is essentially a duplicate of #6851. Closing.
